library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java

```diff
@@ -1407,23 +1408,45 @@ private boolean readSample(ExtractorInput input) throws IOException {
           processSeiNalUnitPayload =
               ceaTrackOutputs.length > 0
                   && NalUnitUtil.isNalUnitSei(track.format.sampleMimeType, nalPrefixData[4]);
+          processSpsNalUnitPayload =
+              NalUnitUtil.isNalUnitSps(track.format.sampleMimeType, nalPrefixData[4]);
           sampleBytesWritten += 5;
           sampleSize += nalUnitLengthFieldLengthDiff;
         } else {
           int writtenBytes;
-          if (processSeiNalUnitPayload) {
-            // Read and write the payload of the SEI NAL unit.
+          if (processSeiNalUnitPayload || processSpsNalUnitPayload) {
+            // We need to process the payload further, so read it into the temporary buffer
+            // before writing.
             nalBuffer.reset(sampleCurrentNalBytesRemaining);
             input.readFully(nalBuffer.getData(), 0, sampleCurrentNalBytesRemaining);
             output.sampleData(nalBuffer, sampleCurrentNalBytesRemaining);
             writtenBytes = sampleCurrentNalBytesRemaining;
-            // Unescape and process the SEI NAL unit.
-            int unescapedLength =
-                NalUnitUtil.unescapeStream(nalBuffer.getData(), nalBuffer.limit());
-            // If the format is H.265/HEVC the NAL unit header has two bytes so skip one more byte.
-            nalBuffer.setPosition(MimeTypes.VIDEO_H265.equals(track.format.sampleMimeType) ? 1 : 0);
-            nalBuffer.setLimit(unescapedLength);
-            CeaUtil.consume(sampleTimeUs, nalBuffer, ceaTrackOutputs);
+            if (processSeiNalUnitPayload) {
+              // Unescape and process the SEI NAL unit.
+              int unescapedLength =
+                  NalUnitUtil.unescapeStream(nalBuffer.getData(), nalBuffer.limit());
```

Comment:
For SEI NAL unit handling, the data is unescaped before being processed. I'm not sure, but might this be necessary for the SPS case as well?

Reply:
I just tested it and unescaping the SPS NALUs too does at least not break things. But I had no luck yet finding any documentation regarding what this unescaping is for and when it's needed. Could you point me to a term or chapter in the ISO standard(s)?